server/blob/blob-storage-strategy/src/main/java/org/apache/james/server/blob/deduplication/BloomFilterGCAlgorithm.java

```diff
@@ -0,0 +1,212 @@
+            this.gcedBlobCount = new AtomicLong();
+            this.errorCount = new AtomicLong();
+            this.bloomFilterExpectedBlobCount = bloomFilterExpectedBlobCount;
+            this.bloomFilterAssociatedProbability = bloomFilterAssociatedProbability;
+        }
+
+        public void incrementBlobCount() {
+            blobCount.incrementAndGet();
+        }
+
+        public void incrementReferenceSourceCount() {
+            referenceSourceCount.incrementAndGet();
+        }
+
+        public void incrementGCedBlobCount() {
+            gcedBlobCount.incrementAndGet();
+        }
+
+        public void incrementErrorCount() {
+            errorCount.incrementAndGet();
+        }
+
+        Snapshot snapshot() {
+            return new Snapshot(
+                referenceSourceCount.get(),
+                blobCount.get(),
+                gcedBlobCount.get(),
+                errorCount.get(),
+                bloomFilterExpectedBlobCount,
+                bloomFilterAssociatedProbability);
+        }
+    }
+
+    private final BlobReferenceSource referenceSource;
+    private final BlobStoreDAO blobStoreDAO;
+    private final GenerationAwareBlobId.Factory generationAwareBlobIdFactory;
+    private final GenerationAwareBlobId.Configuration generationAwareBlobIdConfiguration;
+    private final Instant now;
+
+    // Avoids two subsequent run to have the same false positives.
+    private final String salt;
+
+    public BloomFilterGCAlgorithm(BlobReferenceSource referenceSource,
+                                  BlobStoreDAO blobStoreDAO,
+                                  GenerationAwareBlobId.Factory generationAwareBlobIdFactory,
+                                  GenerationAwareBlobId.Configuration generationAwareBlobIdConfiguration,
+                                  Clock clock) {
+        this.referenceSource = referenceSource;
+        this.blobStoreDAO = blobStoreDAO;
+        this.generationAwareBlobIdFactory = generationAwareBlobIdFactory;
+        this.generationAwareBlobIdConfiguration = generationAwareBlobIdConfiguration;
+        this.salt = UUID.randomUUID().toString();
+        this.now = clock.instant();
+    }
+
+    public Mono<Result> gc(int expectedBlobCount, double associatedProbability, BucketName bucketName, Context context) {
+        return populatedBloomFilter(expectedBlobCount, associatedProbability, context)
+            .flatMap(bloomFilter -> gc(bloomFilter, bucketName, context));
+    }
+
+    public Mono<Result> gc(BloomFilter<CharSequence> bloomFilter, BucketName bucketName, Context context) {
+        return Flux.from(blobStoreDAO.listBlobs(bucketName))
+            .doOnNext(blobId -> context.incrementBlobCount())
+            .flatMap(blobId -> gcBlob(bloomFilter, blobId, bucketName, context))
+            .reduce(Task::combine);
+    }
+
+    public Mono<BloomFilter<CharSequence>> populatedBloomFilter(int expectedBlobCount, double associatedProbability, Context context) {
```

Comment:
private

Reply:
It should be `public`, because it is called in BlobGcService. 
Build one-time bloomFilter for multi bucketNames

Reply:
For now we should just run the GC on the default bucket.